### PR TITLE
NOV-1384: Remove trailing authorization headers from OAuth token

### PR DESCRIPTION
**Pull request by mattiasrosberg**

### CODE REVIEW

File: dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy [hunk 1]
```diff
@@ -273,4 +275,26 @@ class SpringSecurityAuthenticatorSpec extends Specification {
         actual.get().additionalInformation == ["principal": "device:1234", "exp": 3025890541]
     }
 
+    def 'token with trailing authorization headers are cleaned correctly'() {
```
Comment (jeff-blaisdell) [inline dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy]: Wouldn't you need an `@Unroll` on this method in order to leverage the `where` block?

File: dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy [hunk 2]
```diff
@@ -273,4 +275,26 @@ class SpringSecurityAuthenticatorSpec extends Specification {
...
+        def cleanToken = springSecurityAuthenticator.removeTrailingAuthorizationHeaders(tokenValue)
+
+        then:
+        cleanToken == result
+        noExceptionThrown()
+
+        where:
+
+        tokenValue      | result
+        token + ", Signature keyId=\"/SmartThings/33:0a:6e:17:da:b2:2c:ee:63:6c:ee:9e:62:9b:07" +
```
Comment (jeff-blaisdell) [inline dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy]: This is a bit odd.

Is the goal to send both a standard Bearer token and HTTP Signature in 1 Authorization header?

**Comment by jeff-blaisdell** [comment]
I can't say I fully understand what's trying to be accomplished here, but in reading the JIRA ticket description - it seems you are trying to accomplish something non-standard.

It looks like you're trying to truncate the Authorization header value based on a comma character.  If so, I don't think we want that to be standard functionality of our OSS auth library.

File: dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy (hunk 2 above)
Comment (mattiasrosberg) [reply dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy]: Not in the long run but in a transition period we would like to be able to move from OAuth to HTTP Signature with zero downtime. This means we need to send both headers until the endpoint supports both ways of authorization. Then the OAuth can be removed from the Authorization header.

Comment (jeff-blaisdell) [reply dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy]: The dropwizard-oauth currently supports the ability to override it's default behavior with something custom.

The `SpringSecurityAuthenticator` your changes are in are defined by Guice as an Optional binding.  Meaning you can make your own authenticator and override it's injection into the auth mechanisms.

https://github.com/SmartThingsOSS/dropwizard-common/blob/master/dropwizard-oauth/src/main/java/smartthings/dw/oauth/AuthModule.java#L22

I'd suggest closing this PR, and overriding that class in your local project to do whatever custom logic your team needs.

**Comment by mattiasrosberg** [comment]
The root problem, as I see it, is in Dropwizard-auth, OAuthCredentialAuthFilter. https://github.com/dropwizard/dropwizard/blob/master/dropwizard-auth/src/main/java/io/dropwizard/auth/oauth/OAuthCredentialAuthFilter.java#L50

The getCredentials method is assuming a single HTTP header value in the Authorization header. According to the HTTP Standard multiple values are allowed in HTTP headers as a comma separated list. The Dropwizard implementation of getCredentials does not take this into account and simply removed a prefix, in the OAuth case 'Bearer', from a Authorization header value. This is why we get problems in SmartThings dropwizard-common when using multiple Authorization Headers.

In my opinion the best solution would have been to solve this in Dropwizard by looking at all possible values in the Authorization header. Then select the one starting with the string equal to the specified prefix (Bearer) and let that value be the credentials. That would have been clean, standard and simple.

My code is basically just trying to fix the fact that the dropwizard code can't handle multiple http header values in the Authorization header.

File: dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy (hunk 1 above)
Comment (mattiasrosberg) [reply dropwizard-oauth/src/test/groovy/smartthings/dw/oauth/SpringSecurityAuthenticatorSpec.groovy]: No, the @Unroll will only make each row in the where-block be treated as a separate test case. Now all the rows will be treated as a single test case.

**Comment by mattiasrosberg** [comment]
I'm going to implement this in Marconi instead. This way we don't need to do anything non-standard here.

**Pull request closed without merging**